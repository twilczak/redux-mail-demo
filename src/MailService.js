import axios from 'axios';

export class MailService {

    static hostUrl = 'http://localhost:3000';

    static getMessages(mailbox) {
        const url = `${this.hostUrl}/` + mailbox.split('/').join('');
        return axios
            .get(url)
            .then((response) => response.data);
    }

    static getMessage(mailbox, id) {
        const url = `${this.hostUrl}/${mailbox}`;
        return axios
            .get(url)
            .then(response => {
                return response.data.find(message => message.id === id);
            });
    }

    static deleteMessage(mailbox, id) {
        const url = `${this.hostUrl}/${mailbox}/${id}`;
        return axios
            .delete(url)
            .then((response) => response.data);
    }

    static sendMessage(message) {
        const url = `${this.hostUrl}/outbox`;
        message.dateSent = this.getDateSent();

        return axios
            .post(url, message)
            .then(response => response.data);
    }

    static zeroPad(value) {
        return value > 9 ?  '' + value : '0' + value;
    }

    static getDateSent(date = new Date()) {
        const day = this.zeroPad(date.getDate());
        const month = this.zeroPad(date.getMonth() + 1);
        const year = date.getFullYear();

        return `${year}.${month}.${day}`;
    }
}