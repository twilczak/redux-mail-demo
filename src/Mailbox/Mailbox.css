.mailbox {
  display: grid;
  grid-template-columns: 160px auto;
  padding: 10px; }

.mailbox-border {
  width: 100%;
  border: 0 solid black;
  border-bottom-width: 1px;
  grid-column: 1/3; }
