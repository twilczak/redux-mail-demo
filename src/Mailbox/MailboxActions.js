import 'whatwg-fetch';
import { hostUrl } from '../constants';

export const FETCH_MESSAGES = 'FETCH_MESSAGES';
export const FETCH_MESSAGES_SUCCESS = 'FETCH_MESSAGES_SUCCESS';
export const FETCH_MESSAGES_FAILED = 'FETCH_MESSAGES_FAILED';

export const mailboxActions = {
    FETCH_MESSAGES,
    FETCH_MESSAGES_FAILED,
    FETCH_MESSAGES_SUCCESS,
};

export const fetchMessages = (mailbox) => dispatch => {
    const parsedMailbox =  mailbox.split('/').join('');
    const url = `${hostUrl}/${parsedMailbox}`;

    dispatch({type: FETCH_MESSAGES, mailbox: parsedMailbox});

    fetch(url)
        .then(response => response.json())
        .then(jsonResponse => {
            dispatch({type: FETCH_MESSAGES_SUCCESS, mailbox: parsedMailbox, messages: jsonResponse});
        })
        .catch(error => {
            dispatch({type: FETCH_MESSAGES_FAILED, mailbox: parsedMailbox, error});
        });
};

