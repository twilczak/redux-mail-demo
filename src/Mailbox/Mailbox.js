import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';

import './Mailbox.css';

import { MailService } from "../MailService";
import { errorHandler } from "../errorHandler";

import { MailboxControls } from "./MailboxControls/MailboxControls";
import { MailboxList } from "./MailboxList/MailboxList";
import { MessageReader } from "../MessageReader/MessageReader";
import { MessageComposer } from "../MessageComposer/MessageComposer";

export class Mailbox extends Component {

    constructor(props) {
        super(props);

        this.state = {
            messages: []
        };
    }

    componentDidMount() {
        this.loadMessages(this.props.match.path);
    }

    componentWillReceiveProps({match}) {
        this.loadMessages(match.path);
    }

    loadMessages(path) {
        this.setState({messages: []});
        MailService
            .getMessages(path)
            .then((messages) => this.setState({messages: messages}))
            .catch(errorHandler);
    }

    render() {
        return (
            <section className="mailbox">

                <MailboxControls match={this.props.match}/>
                <MailboxList match={this.props.match} location={this.props.location} messages={this.state.messages}/>

                <hr className="mailbox-border"/>

                <Switch>
                    <Route exact path="/inbox/view/:messageId" component={MessageReader}/>
                    <Route exact path="/outbox/view/:messageId" component={MessageReader}/>

                    <Route exact path="/inbox/compose" component={MessageComposer}/>
                    <Route exact path="/outbox/compose" component={MessageComposer}/>
                </Switch>

            </section>
        );
    }
}