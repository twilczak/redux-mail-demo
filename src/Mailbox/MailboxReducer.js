import { mailboxActions } from './MailboxActions';
import { messageReaderActions } from '../MessageReader/MessageReaderActions';
import { messageComposerActions } from '../MessageComposer/MessageComposerActions';

export const mailbox = (state = {isLoading: false, inbox: [], outbox: []}, action) => {
    switch(action.type) {
        case mailboxActions.FETCH_MESSAGES:
            return {
                ...state,
                isLoading: true,
                error: null,
            };
        case mailboxActions.FETCH_MESSAGES_SUCCESS:
            return {
                ...state,
                [action.mailbox]: action.messages,
                isLoading: false,
            };
        case mailboxActions.FETCH_MESSAGES_FAILED:
            return {
                ...state,
                isLoading: false,
                error: action.error,
            };
        case messageReaderActions.DELETE_MESSAGE_SUCCESS:
            return {
                ...state,
                [action.mailbox] : state[action.mailbox].filter(message => message.id !== action.messageId),
            };
        case messageComposerActions.SEND_MESSAGE_SUCCESS:
            return {
                ...state,
                outbox: [...state.outbox, action.message],
            };
        default:
            return state;
    }
};