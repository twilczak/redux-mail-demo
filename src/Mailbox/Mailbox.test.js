import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { MemoryRouter, Route } from 'react-router-dom';
import { Mailbox } from "./Mailbox";

import { MailService } from "../MailService";

jest.mock('../MailService');

describe(Mailbox, () => {
    const mockMessages = [{id: 'test'}];
    const promise = new Promise((resolve, reject) => {
        resolve(mockMessages);
    });

    MailService.getMessages.mockReturnValue(promise);

    const component = shallow(
        <Mailbox match={{path: '/inbox'}}/>
    );

    it('renders and matches the snapshot', () => {
        const component = renderer.create(
            <MemoryRouter initialEntries={['/inbox']}>
                <Route path="/inbox" component={Mailbox}/>
            </MemoryRouter>
        );
        expect(component).toMatchSnapshot();
    });

    it('renders the mailbox markup', () => {
        expect(component.find('.mailbox')).toBeTruthy();
        expect(component.find('MailboxControls')).toBeTruthy();
        expect(component.find('MailboxList')).toBeTruthy();
        expect(component.find('.mailbox-border')).toBeTruthy();
    });

    it('calls MailService.getMessages()', () => {
        expect(MailService.getMessages).toBeCalledWith('/inbox');
        expect(component.state()).toEqual({messages: [{id: 'test'}]});
    });
});
