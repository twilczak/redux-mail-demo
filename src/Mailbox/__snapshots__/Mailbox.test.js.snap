// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Mailbox extends _react.Component {

    constructor(props) {
        super(props);

        this.state = {
            messages: [] };

    }

    componentDidMount() {
        this.loadMessages(this.props.match.path);
    }

    componentWillReceiveProps({ match }) {
        this.loadMessages(match.path);
    }

    loadMessages(path) {
        this.setState({ messages: [] });
        _MailService.MailService.
        getMessages(path).
        then(messages => this.setState({ messages: messages })).
        catch(_errorHandler.errorHandler);
    }

    render() {
        return (
            _react2.default.createElement('section', { className: 'mailbox', __source: { fileName: _jsxFileName, lineNumber: 42 }, __self: this },

                _react2.default.createElement(_MailboxControls.MailboxControls, { match: this.props.match, __source: { fileName: _jsxFileName, lineNumber: 44 }, __self: this }),
                _react2.default.createElement(_MailboxList.MailboxList, { match: this.props.match, location: this.props.location, messages: this.state.messages, __source: { fileName: _jsxFileName, lineNumber: 45 }, __self: this }),

                _react2.default.createElement('hr', { className: 'mailbox-border', __source: { fileName: _jsxFileName, lineNumber: 47 }, __self: this }),

                _react2.default.createElement(_reactRouterDom.Switch, { __source: { fileName: _jsxFileName, lineNumber: 49 }, __self: this },
                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/inbox/view/:messageId', component: _MessageReader.MessageReader, __source: { fileName: _jsxFileName, lineNumber: 50 }, __self: this }),
                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/outbox/view/:messageId', component: _MessageReader.MessageReader, __source: { fileName: _jsxFileName, lineNumber: 51 }, __self: this }),

                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/inbox/compose', component: _MessageComposer.MessageComposer, __source: { fileName: _jsxFileName, lineNumber: 53 }, __self: this }),
                    _react2.default.createElement(_reactRouterDom.Route, { exact: true, path: '/outbox/compose', component: _MessageComposer.MessageComposer, __source: { fileName: _jsxFileName, lineNumber: 54 }, __self: this }))));




    }} renders and matches the snapshot 1`] = `
<section
  className="mailbox"
>
  <div>
    <a
      aria-current="true"
      className="mailbox-button is-active"
      href="/inbox"
      onClick={[Function]}
      style={Object {}}
    >
      Inbox
    </a>
    <a
      aria-current={false}
      className="mailbox-button"
      href="/outbox"
      onClick={[Function]}
      style={undefined}
    >
      Outbox
    </a>
    <a
      className="mailbox-button"
      href="/inbox/compose"
      onClick={[Function]}
    >
      Compose
    </a>
  </div>
  <ol
    className="mailbox-list"
  >
    <li
      className="mailbox-list-header"
    >
      <span
        className="sender"
      >
        Sender
      </span>
      <span
        className="subject"
      >
        Subject
      </span>
      <span
        className="date-sent"
      >
        Date
      </span>
    </li>
  </ol>
  <hr
    className="mailbox-border"
  />
</section>
`;
